databuilder/query_engines/base_sql.py

```diff
@@ -240,7 +276,39 @@ def get_sql_element_or_value(self, value: Any) -> Any:
         themselves (which require resolving to SQL) or plain static values (booleans,
         integers, dates, list of dates etc) which we can return unchanged to SQLAlchemy.
         """
-        if isinstance(value, QueryNode):
+        types_accepted_by_get_sql_element_statically = {
+            FilteredTable,
+            Row,
+            RowFromAggregate,
+            ReifiedQuery,
+            ColumnSelectorNode,
+            ValueFromCategory,
+            Comparator,
+            Codelist,
+            ValueFromFunction,
+        }
+
+        types_accepted_by_get_sql_element_dynamically = set()
+
+        register_method = BaseSQLQueryEngine.get_sql_element_no_cache.register
+        # mypy thinks this is a just Callable, not a bound method
+        singledispatchmethod_obj = register_method.__self__  # type: ignore[attr-defined]
+        for function in singledispatchmethod_obj.dispatcher.registry.values():
+            signature = function.__annotations__
+            arg_keys = list(signature.keys())
+            arg_keys.remove("return")
+            assert (
+                len(arg_keys) == 1
+            ), "Expect instances of get_sql_element_no_cache() to have only one argument"
+            arg_type = signature[arg_keys[0]]
+            types_accepted_by_get_sql_element_dynamically.add(arg_type)
+
+        assert (
+            types_accepted_by_get_sql_element_dynamically
+            == types_accepted_by_get_sql_element_statically
+        )
+
+        if type(value) in types_accepted_by_get_sql_element_statically:
```

Comment:
We need the real list of types here rather than just `QueryNode` because otherwise the type checker doesn't know which branch we're going down.